### PR TITLE
[JCLOUDS-1430] Aliyun ECS initial skeleton

### PR DESCRIPTION
**Pull request by andreaturli**
- add ApiMetadata and ProviderMetadata
- add skeleton for ECSComputeApi with ImageApi only
- add `FormSign` filter
- add HttpApiModule and ParserModule
- add `ImageApi` feature with Mock and Live Tests

### CODE REVIEW

File: aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/domain/options/ListImagesOptions.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+      queryParameters.put(IMAGE_OWNER_ALIAS_PARAM, imageOwnerAlias);
+      return this;
+   }
+
+   public ListImagesOptions usage(String usage) {
+      queryParameters.put(USAGE_PARAM, usage);
+      return this;
+   }
+
+   public ListImagesOptions paginationOptions(final PaginationOptions paginationOptions) {
```
Comment (nacx) [inline aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/domain/options/ListImagesOptions.java]: Could it be better for this class to extend `PaginationOptions` instead?

File: aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/domain/options/PaginationOptions.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import static com.google.common.base.Preconditions.checkState;
+
+public class PaginationOptions extends BaseHttpRequestOptions {
+
+   public static final String PAGE_NUMBER = "pageNumber";
+   public static final String PAGE_SIZE = "pageSize";
+
+   public PaginationOptions pageNumber(int pageNumber) {
+      checkState(pageNumber > 0, "pageSize must be > 0");
+      checkState(pageNumber <= 10000, "limit must be <= 50");
```
Comment (nacx) [inline aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/domain/options/PaginationOptions.java]: Precondition and message don't match.

File: aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/domain/options/PaginationOptions.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+import static com.google.common.base.Preconditions.checkState;
+
+public class PaginationOptions extends BaseHttpRequestOptions {
+
+   public static final String PAGE_NUMBER = "pageNumber";
+   public static final String PAGE_SIZE = "pageSize";
+
+   public PaginationOptions pageNumber(int pageNumber) {
+      checkState(pageNumber > 0, "pageSize must be > 0");
```
Comment (nacx) [inline aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/domain/options/PaginationOptions.java]: Wrong message

File: aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/features/ImageApi.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.MediaType;
+
+/**
+ * https://www.alibabacloud.com/help/doc-detail/25534.htm?spm=a2c63.p38356.b99.330.79eb59abhmnMDE
+ */
+@Consumes(MediaType.APPLICATION_JSON)
+@RequestFilters(FormSign.class)
+@QueryParams(keys = {"Version", "Format", "SignatureVersion", "ServiceCode", "SignatureMethod"},
+        values = {"2014-05-26", "JSON", "1.0", "ecs", "HMAC-SHA1"})
```
Comment (nacx) [inline aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/features/ImageApi.java]: Is this date the API version that can be set in the `ContextBuilder`? In that case instead of hardcoding you can resolve its value. Look at the Chef API for an example.

File: aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/features/ImageApi.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+public interface ImageApi {
+
+   @Named("image:list")
+   @GET
+   @QueryParams(keys = "Action", values = "DescribeImages")
+   @ResponseParser(ParseImages.class)
+   @Fallback(Fallbacks.EmptyIterableWithMarkerOnNotFoundOr404.class)
+   PaginatedCollection<Image> list(@QueryParam("RegionId") String region, ListImagesOptions... options);
```
Comment (nacx) [inline aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/features/ImageApi.java]: Don't do this, as it is not OK to allow passing N option objects. Provide a method without options, and an overloaded one with just one options parameter.

File: aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/features/ImageApi.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+   final class ParseImages extends ParseJson<Images> {
+
+      @Inject
+      ParseImages(final Json json) {
+         super(json, TypeLiteral.get(Images.class));
+      }
+
+      static class ToPagedIterable extends Arg0ToPagedIterable<Image, ToPagedIterable> {
+
+         private ECSComputeServiceApi api;
```
Comment (nacx) [inline aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/features/ImageApi.java]: Make it final

File: aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/filters/AcsURLEncoder.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jclouds.aliyun.ecs.filters;
+
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+
+public class AcsURLEncoder {
```
Comment (nacx) [inline aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/filters/AcsURLEncoder.java]: Can't you just use [Strings2#urlEncode](https://github.com/jclouds/jclouds/blob/master/core/src/main/java/org/jclouds/util/Strings2.java#L56)? Or do you need to specifically handle the `~` characterd?

File: aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/filters/FormSign.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+      } catch (UnsupportedEncodingException e) {
+         throw Throwables.propagate(e);
+      }
+
+      // encode each parameter value first,
+      String stringToSign = prefix;
+      ImmutableSortedSet.Builder<String> builder = ImmutableSortedSet.naturalOrder();
+      for (Map.Entry<String, String> entry : decodedParams.entries())
+         builder.add(Strings2.urlEncode(entry.getKey()) + "=" + Strings2.urlEncode(entry.getValue()));
+      stringToSign += Strings2.urlEncode(Joiner.on("&").join(builder.build()));
```
Comment (nacx) [inline aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/filters/FormSign.java]: If you use the jclouds `Uris` and `UriBuilder` classes I think you don't have to worry about manually encoding the parameters.

File: aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/filters/FormSign.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+   public HttpRequest filter(HttpRequest request) throws HttpException {
+      Credentials currentCreds = checkNotNull(creds.get(), "credential supplier returned null");
+
+      Signer signer = Signer.getSigner();
+
+      String signature;
+
+      Multimap<String, String> decodedParams = queryParser().apply(request.getEndpoint().getRawQuery());
+
+      SimpleDateFormat df = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'");
```
Comment (nacx) [inline aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/filters/FormSign.java]: This could be declared as a constant.

File: aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/filters/HmacSHA1Signer.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+public class HmacSHA1Signer extends Signer {
+
+   private static final String ALGORITHM_NAME = "HmacSHA1";
+
+   @Override
+   public String signString(String stringToSign, String accessKeySecret) {
+      try {
+         Mac mac = Mac.getInstance(ALGORITHM_NAME);
+         mac.init(new SecretKeySpec(accessKeySecret.getBytes(Charsets.UTF_8), ALGORITHM_NAME));
+         byte[] signData = mac.doFinal(stringToSign.getBytes(Charsets.UTF_8));
```
Comment (nacx) [inline aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/filters/HmacSHA1Signer.java]: Instead of manually encrypting, get injected a [Crypto](https://github.com/jclouds/jclouds/blob/master/core/src/main/java/org/jclouds/crypto/Crypto.java#L46) object and use it. This way users will be able to configure they preferred security provider, such as Bouncycastle.

File: aliyun-ecs/src/test/java/org/jclouds/aliyun/ecs/compute/internal/BaseECSComputeServiceApiLiveTest.java [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+   protected void assertNodeSuspended(String instanceId) {
+      assertTrue(instanceSuspended.apply(instanceId),
+            String.format("Instance %s was not suspended in the configured timeout", instanceId));
+   }
+
+   protected void assertNodeTerminated(String instanceId) {
+      assertTrue(instanceTerminated.apply(instanceId),
+            String.format("Instance %s was not terminated in the configured timeout", instanceId));
+   }
```
Comment (nacx) [inline aliyun-ecs/src/test/java/org/jclouds/aliyun/ecs/compute/internal/BaseECSComputeServiceApiLiveTest.java]: Remove these methods until needed.

File: aliyun-ecs/src/test/java/org/jclouds/aliyun/ecs/compute/internal/BaseECSComputeServiceApiMockTest.java [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+   }
+
+   protected RecordedRequest assertSent(MockWebServer server, String method, String path, String json)
+         throws InterruptedException {
+      RecordedRequest request = assertSent(server, method, path);
+      assertEquals(request.getHeader("Content-Type"), "application/json");
+      assertEquals(parser.parse(new String(request.getBody(), Charsets.UTF_8)), parser.parse(json));
+      return request;
+   }
+}
```
Comment (nacx) [inline aliyun-ecs/src/test/java/org/jclouds/aliyun/ecs/compute/internal/BaseECSComputeServiceApiMockTest.java]: Remove all methods that are not used.

File: aliyun-ecs/src/test/java/org/jclouds/aliyun/ecs/compute/features/ImageApiMockTest.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+      server.enqueue(jsonResponse("/images-first.json"));
+      server.enqueue(jsonResponse("/images-second.json"));
+      server.enqueue(jsonResponse("/images-last.json"));
+
+      Iterable<Image> images = api.imageApi().list(Regions.EU_CENTRAL_1.getName()).concat();
+      assertEquals(size(images), 28); // Force the PagedIterable to advance
+      assertEquals(server.getRequestCount(), 3);
+      assertSent(server, "GET", "DescribeImages");
+      assertSent(server, "GET", "DescribeImages", 2);
+      assertSent(server, "GET", "DescribeImages", 3);
```
Comment (nacx) [inline aliyun-ecs/src/test/java/org/jclouds/aliyun/ecs/compute/features/ImageApiMockTest.java]: We should also verify that we're sending the page size parameter.

Comment (nacx) [reply aliyun-ecs/src/test/java/org/jclouds/aliyun/ecs/compute/features/ImageApiMockTest.java]: And to leave the pagination stuff properly tested, add a test for a single page response, and for an empty list response.

File: aliyun-ecs/pom.xml [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.apache.jclouds.labs</groupId>
+        <artifactId>jclouds-labs</artifactId>
+        <version>2.2.0-SNAPSHOT</version>
+    </parent>
+    <artifactId>aliyun-ecs</artifactId>
+    <name>Cloudsoft :: jclouds Alibaba Elastic Compute Service API</name>
```
[the remark this answers is not in the document]
Comment (andreaturli) [reply aliyun-ecs/pom.xml]: ops

File: aliyun-ecs/pom.xml [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+                    <instructions>
+                        <!--<Bundle-Activator>${jclouds.osgi.activator}</Bundle-Activator>-->
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <!--<Export-Package>${jclouds.osgi.export}</Export-Package>-->
+                        <Import-Package>${guava.osgi.import},${jclouds.osgi.import}</Import-Package>
+                        <DynamicImport-Package>*</DynamicImport-Package>
+                        <!--<Fragment-Host>${jclouds.osgi.fragment}</Fragment-Host>-->
+                    </instructions>
+                </configuration>
+            </plugin>
```
[the remark this answers is not in the document]
Comment (andreaturli) [reply aliyun-ecs/pom.xml]: removing

File: aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/ECSServiceApiMetadata.java [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+   }
+
+   protected ECSServiceApiMetadata(Builder builder) {
+      super(builder);
+   }
+
+   public static Properties defaultProperties() {
+      Properties properties = BaseHttpApiMetadata.defaultProperties();
+      properties.put(TEMPLATE,
+              "osFamily=CENTOS,os64Bit=true,osVersionMatches=7.4,hardwareId=ecs.t5-lc2m1.nano,locationId=eu-central-1");
```
[the remark this answers is not in the document]
Comment (andreaturli) [reply aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/ECSServiceApiMetadata.java]: ok

File: aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/config/ECSComputeServiceHttpApiModule.java [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+import org.jclouds.rest.ConfiguresHttpApi;
+import org.jclouds.rest.config.HttpApiModule;
+
+@ConfiguresHttpApi
+public class ECSComputeServiceHttpApiModule extends HttpApiModule<ECSComputeServiceApi> {
+
+   @Override
+   protected void configure() {
+      super.configure();
+      bind(ImplicitLocationSupplier.class).to(FirstRegion.class).in(Scopes.SINGLETON);
```
[the remark this answers is not in the document]
Comment (andreaturli) [reply aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/config/ECSComputeServiceHttpApiModule.java]: I'll use that and I'll double check later

File: aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/domain/Image.java [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+
+   public abstract Boolean isSelfShared();
+
+   public abstract Boolean isCopied();
+
+   public abstract Boolean isSubscribed();
+
+   public abstract String platform();
+
+   public abstract String size();
```
[the remark this answers is not in the document]
Comment (andreaturli) [reply aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/domain/Image.java]: not clear yet, I'll keep it in mind. The builder could be useful to create images but I've not implemented it yet

File: aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/domain/Regions.java [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+    * @return Region enum representing the given region name.
+    */
+   public static Regions fromName(String regionName) {
+      for (Regions region : Regions.values()) {
+         if (region.getName().equals(regionName)) {
+            return region;
+         }
+      }
+      throw new IllegalArgumentException("Cannot create enum from " + regionName + " value!");
+   }
```
[the remark this answers is not in the document]
Comment (andreaturli) [reply aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/domain/Regions.java]: yes, will push in subsequent PRs

File: aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/domain/Regions.java [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+   CN_NORTH_1("cn-qingdao", "China (Qingdao)"),
+   CN_NORTH_2("cn-beijing", "China (Beijing)"),
+   CN_NORTH_3("cn-zhangjiakou", "China (Zhangjiakou)"),
+   CN_NORTH_5("cn-huhehaote", "China (Huhehaote)"),
+   CN_EAST_1("cn-hangzhou", "China (Zhangjiakou)"),
+   CN_EAST_2("cn-shanghai", "China (Shanghai)"),
+   CN_SOUTH_1("cn-shenzhen", "China (Shenzhen)"),
+   CN_SOUTH_2("cn-hongkong", "China (Hongkong)"),
+   CN_NORTHWEST_1("cn-northwest-1", "India (Mumbai)"),
+   ME_EAST_1("me-east-1", "UAE (Dubai)");
```
[the remark this answers is not in the document]
Comment (andreaturli) [reply aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/domain/Regions.java]: `Regions` enum is only a util to be used in tests, not sure it is really needed. I'll keep it for now

File: aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/domain/options/PaginationOptions.java (hunk 2 above)
Comment (andreaturli) [reply aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/domain/options/PaginationOptions.java]: ops

File: aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/domain/options/PaginationOptions.java (hunk 3 above)
Comment (andreaturli) [reply aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/domain/options/PaginationOptions.java]: thx

File: aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/features/ImageApi.java (hunk 4 above)
Comment (andreaturli) [reply aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/features/ImageApi.java]: ok sounds better

File: aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/filters/AcsURLEncoder.java (hunk 7 above)
Comment (andreaturli) [reply aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/filters/AcsURLEncoder.java]: thanks I think I missed that, not sure I need `~`

File: aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/features/ImageApi.java [hunk 21]
```diff
@@ -52,15 +52,15 @@
 @Consumes(MediaType.APPLICATION_JSON)
 @RequestFilters(FormSign.class)
 @QueryParams(keys = {"Version", "Format", "SignatureVersion", "ServiceCode", "SignatureMethod"},
-        values = {"2014-05-26", "JSON", "1.0", "ecs", "HMAC-SHA1"})
+        values = { ECSServiceApiMetadata.DEFAULT_API_VERSION, "JSON", "1.0", "ecs", "HMAC-SHA1"})
```
Comment (nacx) [inline aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/features/ImageApi.java]: Use `"{" + Constants.PROPERTY_API_VERSION + "}"` instead for the version, so the value of the variable is used. This way instead of having a hardcoded value, it will pick the default one, or the one configured by users at runtime when creating the context via the `apiVersion` method.

File: aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/features/ImageApi.java [hunk 22]
```diff
@@ -52,15 +52,15 @@
...
 public interface ImageApi {
 
    @Named("image:list")
    @GET
    @QueryParams(keys = "Action", values = "DescribeImages")
    @ResponseParser(ParseImages.class)
    @Fallback(Fallbacks.EmptyIterableWithMarkerOnNotFoundOr404.class)
-   PaginatedCollection<Image> list(@QueryParam("RegionId") String region, ListImagesOptions... options);
+   IterableWithMarker<Image> list(@QueryParam("RegionId") String region, ListImagesOptions options);
```
Comment (nacx) [inline aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/features/ImageApi.java]: It is OK to return PaginatedCollection

File: aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/filters/FormSign.java [hunk 23]
```diff
@@ -30,39 +31,44 @@
...
 
-      Signer signer = Signer.getSigner();
+      Multimap<String, String> decodedParams = queryParser().apply(request.getEndpoint().getQuery());
 
-      String signature;
-
-      Multimap<String, String> decodedParams = queryParser().apply(request.getEndpoint().getRawQuery());
-
-      SimpleDateFormat df = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'");
+      SimpleDateFormat df = new SimpleDateFormat(ECS_DATE_FORMAT);
```
Comment (nacx) [inline aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/filters/FormSign.java]: Why not declare the `SimpleDateFormat` object as a constant?

File: aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/features/ImageApi.java (hunk 21 above)
Comment (andreaturli) [reply aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/features/ImageApi.java]: cool, sorry I forgot to address your previous comment on that and I left an intermediate solution

**Comment by andreaturli** [comment]
merged at [master](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/5bd2a80f)
